File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/data/analytics/Logger.kt
package com.example.mobile_android.data.analytics

import android.util.Log
import java.io.File
import java.io.FileWriter
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

enum class LogLevel {
    VERBOSE, DEBUG, INFO, WARN, ERROR
}

data class LogEntry(
    val level: LogLevel,
    val tag: String,
    val message: String,
    val throwable: Throwable? = null,
    val timestamp: Long = System.currentTimeMillis()
)

@Singleton
class Logger @Inject constructor() {
    
    private val logs = mutableListOf<LogEntry>()
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
    
    companion object {
        private const val MAX_LOGS = 1000
    }
    
    /**
     * Log verbose message
     */
    fun v(tag: String, message: String) {
        log(LogLevel.VERBOSE, tag, message)
        Log.v(tag, message)
    }
    
    /**
     * Log debug message
     */
    fun d(tag: String, message: String) {
        log(LogLevel.DEBUG, tag, message)
        Log.d(tag, message)
    }
    
    /**
     * Log info message
     */
    fun i(tag: String, message: String) {
        log(LogLevel.INFO, tag, message)
        Log.i(tag, message)
    }
    
    /**
     * Log warning message
     */
    fun w(tag: String, message: String, throwable: Throwable? = null) {
        log(LogLevel.WARN, tag, message, throwable)
        if (throwable != null) {
            Log.w(tag, message, throwable)
        } else {
            Log.w(tag, message)
        }
    }
    
    /**
     * Log error message
     */
    fun e(tag: String, message: String, throwable: Throwable? = null) {
        log(LogLevel.ERROR, tag, message, throwable)
        if (throwable != null) {
            Log.e(tag, message, throwable)
        } else {
            Log.e(tag, message)
        }
    }
    
    /**
     * Internal log storage
     */
    private fun log(level: LogLevel, tag: String, message: String, throwable: Throwable? = null) {
        val entry = LogEntry(level, tag, message, throwable)
        logs.add(entry)
        
        // Keep only last MAX_LOGS entries
        if (logs.size > MAX_LOGS) {
            logs.removeAt(0)
        }
    }
    
    /**
     * Get all logs
     */
    fun getLogs(): List<LogEntry> = logs.toList()
    
    /**
     * Get logs by level
     */
    fun getLogsByLevel(level: LogLevel): List<LogEntry> {
        return logs.filter { it.level == level }
    }
    
    /**
     * Get logs by tag
     */
    fun getLogsByTag(tag: String): List<LogEntry> {
        return logs.filter { it.tag == tag }
    }
    
    /**
     * Export logs to file
     */
    fun exportLogs(file: File): Boolean {
        return try {
            FileWriter(file).use { writer ->
                logs.forEach { entry ->
                    val timestamp = dateFormat.format(Date(entry.timestamp))
                    val line = "[$timestamp] [${entry.level}] [${entry.tag}] ${entry.message}"
                    writer.appendLine(line)
                    
                    entry.throwable?.let { throwable ->
                        writer.appendLine(throwable.stackTraceToString())
                    }
                }
            }
            true
        } catch (e: Exception) {
            Log.e("Logger", "Failed to export logs", e)
            false
        }
    }
    
    /**
     * Clear all logs
     */
    fun clearLogs() {
        logs.clear()
    }
    
    /**
     * Format log entry as string
     */
    fun formatLogEntry(entry: LogEntry): String {
        val timestamp = dateFormat.format(Date(entry.timestamp))
        return "[$timestamp] [${entry.level}] [${entry.tag}] ${entry.message}"
    }
}

File: external_apps/mobile_ios/mobile_ios/Core/Analytics/Logger.swift
//
//  Logger.swift
//  mobile_ios
//
//  Created for Analytics & Monitoring
//

import Foundation
import os.log

enum LogLevel: String {
    case verbose = "VERBOSE"
    case debug = "DEBUG"
    case info = "INFO"
    case warn = "WARN"
    case error = "ERROR"
}

struct LogEntry {
    let level: LogLevel
    let tag: String
    let message: String
    let error: Error?
    let timestamp: Date
}

class Logger {
    static let shared = Logger()
    
    private var logs: [LogEntry] = []
    private let maxLogs = 1000
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        return formatter
    }()
    
    private init() {}
    
    /// Log verbose message
    func v(_ tag: String, _ message: String) {
        log(.verbose, tag: tag, message: message)
        os_log("%{public}@", log: OSLog(subsystem: tag, category: "verbose"), type: .debug, message)
    }
    
    /// Log debug message
    func d(_ tag: String, _ message: String) {
        log(.debug, tag: tag, message: message)
        os_log("%{public}@", log: OSLog(subsystem: tag, category: "debug"), type: .debug, message)
    }
    
    /// Log info message
    func i(_ tag: String, _ message: String) {
        log(.info, tag: tag, message: message)
        os_log("%{public}@", log: OSLog(subsystem: tag, category: "info"), type: .info, message)
    }
    
    /// Log warning message
    func w(_ tag: String, _ message: String, error: Error? = nil) {
        log(.warn, tag: tag, message: message, error: error)
        os_log("%{public}@", log: OSLog(subsystem: tag, category: "warn"), type: .error, message)
    }
    
    /// Log error message
    func e(_ tag: String, _ message: String, error: Error? = nil) {
        log(.error, tag: tag, message: message, error: error)
        os_log("%{public}@", log: OSLog(subsystem: tag, category: "error"), type: .fault, message)
    }
    
    /// Internal log storage
    private func log(_ level: LogLevel, tag: String, message: String, error: Error? = nil) {
        let entry = LogEntry(level: level, tag: tag, message: message, error: error, timestamp: Date())
        logs.append(entry)
        
        // Keep only last maxLogs entries
        if logs.count > maxLogs {
            logs.removeFirst()
        }
    }
    
    /// Get all logs
    func getLogs() -> [LogEntry] {
        return logs
    }
    
    /// Get logs by level
    func getLogsByLevel(_ level: LogLevel) -> [LogEntry] {
        return logs.filter { $0.level == level }
    }
    
    /// Get logs by tag
    func getLogsByTag(_ tag: String) -> [LogEntry] {
        return logs.filter { $0.tag == tag }
    }
    
    /// Export logs to file
    func exportLogs(to url: URL) -> Bool {
        do {
            var content = ""
            for entry in logs {
                let timestamp = dateFormatter.string(from: entry.timestamp)
                let line = "[\(timestamp)] [\(entry.level.rawValue)] [\(entry.tag)] \(entry.message)\n"
                content += line
                
                if let error = entry.error {
                    content += "\(error.localizedDescription)\n"
                }
            }
            
            try content.write(to: url, atomically: true, encoding: .utf8)
            return true
        } catch {
            print("Logger: Failed to export logs - \(error)")
            return false
        }
    }
    
    /// Clear all logs
    func clearLogs() {
        logs.removeAll()
    }
    
    /// Format log entry as string
    func formatLogEntry(_ entry: LogEntry) -> String {
        let timestamp = dateFormatter.string(from: entry.timestamp)
        return "[\(timestamp)] [\(entry.level.rawValue)] [\(entry.tag)] \(entry.message)"
    }
}
